<img style="float:right" src="images/logo_va.png" /> 

# Visione Artificiale
## Esercitazione: Python e NumPy

### Sommario
* Un primo esercizio per per prendere confidenza con Python: generare i numeri perfetti;
* Un secondo esercizio in Python: calcolare alcune statistiche sulle parole in un testo;
* Un esercizio sullo sli per per prendere confidenza con Python: generare i numeri perfetti;
* Un secondo esercizio in Python: calcolare alcune statistiche sulle parole in un testo;
* Un esercizio sullo slicing di liste in Python;
* Alcune operazioni su matricing di liste in Python;
* Alcune operazioni su matrici in Python e in NumPy;
* Il setaccio di Eratostene con NumPy.

Iniziamo con l'importazione dei moduli che ci serviranno: `NumPy` e `va`.

In [1]:
import numpy as np
import va

<img style="float:left" src="images/ar.png" /> **Esercizio 1** - Nella prossima cella completare la funzione `perfetti(k)` che deve restituire una lista ordinata dei numeri *perfetti* fra 1 e k. Un numero $N$ si dice perfetto se $\sigma(N) = 2\cdot N$, dove la funzione $\sigma(N)$ restituisce la somma dei divisori positivi di $N$. Ad esempio 6 è perfetto in quanto la somma dei suoi divisori (1, 2, 3, 6) è pari a 12.

In [12]:
def perfetti(k):
    rtr = []
    
    def is_perfetto(n):
        divisors = 0
        
        for i in range(1, n + 1):
            if n % i == 0:
                divisors += i
        
        return 2 * n == divisors
    
    for i in range(1, k + 1):
        if is_perfetto(i):
            rtr.append(i)
    
    return rtr
    
    # --- Svolgimento Esercizio 1: Fine --- #

print(perfetti(100))


[6, 28]


Eseguire la cella seguente per controllare se la funzione è corretta.

In [14]:
va.test_basi_1(perfetti)

<img style="float:left" src="images/ar.png" /> **Esercizio 2** - Nella prossima cella completare la funzione `statistiche_parole()` che riceve in input una stringa e deve restituire una tupla contenente i seguenti valori:

* Numero di parole,
* Lunghezza della parola più corta,
* Lunghezza della parola più lunga,
* Lunghezza media delle parole.

Come delimitatori, considerare semplicemente il carattere spazio, virgola e punto.  
Ad esempio, `statistiche_parole("Flat is better than nested.")` deve restituire `(5, 2, 6, 4.4)`.  
Nel caso non ci sia alcuna parola nel testo, la funzione deve restituire `(0, None, None, None)`.  
Suggerimenti: possono essere utili alcuni metodi della classe `str` come `split` e `replace`. 

In [23]:
def statistiche_parole(testo):
    if testo.strip() == "":
        return (0, None, None, None)

    # --- Svolgimento Esercizio 2: Inizio --- #    
    testo_words = testo.split(" ")
    num_word = len(testo_words)
    min_word = testo_words[0]
    max_word = testo_words[0]
    avg = 0.0
    
    for i in testo_words:
        avg += len(i)
        
        if len(i) < len(min_word):
            min_word = i
        
        if len(i) > len(max_word):
            max_word = i
        
    return (num_word, min_word, max_word, avg/num_word)    
    
    # --- Svolgimento Esercizio 2: Fine --- #

Eseguire la cella seguente per verificare la funzione in alcuni casi di prova.

In [24]:
va.test_basi_2(statistiche_parole)

Si consideri ora la seguente lista di interi.

In [ ]:
numeri = list(range(21))
numeri

<img style="float:left" src="images/ar.png" /> **Esercizio 3** - La cella seguente contiene una tupla (`risultati`) con una serie di liste di interi: si tratta di sottoinsiemi della lista `numeri`. Nella cella successiva inizializzare una tupla identica a  utilizzando *solamente* operazioni di slicing sulla lista `numeri`.  
Ad esempio: se fosse `risultati = ([0,1], [0,2,4,6])`, una possibile soluzione sarebbe `slicing = (numeri[:3], numeri[:8:2])`. Nota bene: è richiesto di utilizzare *unicamente* indicizzazione con slicing sulla lista `numeri`: ogni elemento della tupla deve avere la forma `numeri[]` con almeno un ':' all'interno delle parentesi quadre.  
Eseguire la cella successiva per controllare la correttezza di quando fatto.

In [ ]:
risultati = ([0,1,2], [20, 19], [11, 14, 17, 20], [], [1, 19], [19, 1], [11])

In [ ]:
# --- Svolgimento Esercizio 3: Inizio --- #

slicing = () # Costruire la tupla corretta

# --- Svolgimento Esercizio 3: Fine --- #

In [ ]:
va.test_basi_3(risultati, slicing)

Dopo aver svolto alcuni esercizi in Python, nella parte restante di questa esercitazione faremo un po' di pratica con NumPy. Per meglio apprezzare i vantaggi di NumPy, inizieremo con l'analizzare una serie di operazioni su matrici e vettori implementate in puro Python (senza alcuna libreria aggiuntiva), per poi implementarle in NumPy e confrontarne la velocità.

Nella prossima cella è implementata una funzione Python che svolge le seguenti operazioni:

* Creazione di una matrice $A$ con 100 righe e 50 colonne, in cui ogni elemento ha il seguente valore: $A_{i,j}=i\cdot j$, con $(i,j)$ indici di riga e colonna 1-based;
* Creazione di una matrice $B$ con 50 righe e 200 colonne, in cui ogni elemento ha il seguente valore: $B_{i,j}=i+j$, con $(i,j)$ indici di riga e colonna 1-based;
* Calcolare il vettore $cA$ contenente la somma di ciascuna colonna di A;
* Calcolare il vettore $rB$ contenente la somma di ciascuna riga di B;
* Calcolare la matrice $C=A \cdot B$;
* Creare una matrice $D$ a partire da $A$, ponendo pari a 0 tutti i valori divisibili per 7;
* Creare una matrice $E$ a partire da $B$, ponendo pari a -1 tutti i valori il cui indice di riga (1-based) è divisibile per 2 e il cui indice di colonna (1-based) è divisibile per 3.
* Restituisce una tupla contenente tutte le matrici e i vettori creati.

Analizzare e comprendere il codice (eventualmente creando altre celle per fare opportune prove) prima di proseguire.

In [2]:
def operazioni_py():
    A = [ [i * j for j in range(1, 51)] for i in range(1, 101) ]
    B = [ [i + j for j in range(1, 201)] for i in range(1, 51) ]
    cA = [ sum( A[i][j] for i in range(len(A)) ) for j in range(len(A[0]))]
    rB = [ sum(r) for r in B ]
    C = [[sum([A[i][k]*B[k][j] for k in range(len(A[0]))]) for j in range(len(B[0]))] for i in range(len(A))]
    D = [ [0 if a % 7 == 0 else a for a in r] for r in A ]
    E = [ [-1 if ((i+1) % 2 == 0) and ((j+1) % 3 == 0) else B[i][j] for j in range(len(B[0]))] for i in range(len(B)) ]
    return (A, B, C, D, E, cA, rB)

<img style="float:left" src="images/ar.png" /> **Esercizio 4** - Nella cella seguente completate la funzione `operazioni_np()` eseguendo le stesse operazioni di `operazioni_py()` mediante NumPy. In questo caso la tupla restituita dalla funzione dovrà contenere oggetti `ndarray` e non liste Python. Le due celle seguenti confrontano i risultati e misurano i tempi di esecuzione.

In [26]:
def operazioni_np():
    # --- Svolgimento Esercizio 4: Inizio --- #
    
    A, B, cA, rB, C, D, E = [np.zeros(1)] * 7
    
    A = np.fromfunction(lambda x,y: x*y, (100, 50))
    B = np.fromfunction(lambda x,y: x+y, (50, 200))
    cA = A.sum(axis=0)
    cB = B.sum(axis=1)
    C = A @ B
    D = A
    D[A % 7 == 0] = 0
    E = B
    E[1::2, 1::3] = -1
    # --- Svolgimento Esercizio 4: Fine --- #
    
    return (A, B, C, D, E, cA, rB)

In [34]:
Ap, Bp, Cp, Dp, Ep, cAp, rBp = operazioni_py()
print("Python:\n ",Ap)
A = np.fromfunction(lambda x,y: x*y, (100, 50))
print("Numpy\n",A)

Python:
  [[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50], [2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100], [3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 45, 48, 51, 54, 57, 60, 63, 66, 69, 72, 75, 78, 81, 84, 87, 90, 93, 96, 99, 102, 105, 108, 111, 114, 117, 120, 123, 126, 129, 132, 135, 138, 141, 144, 147, 150], [4, 8, 12, 16, 20, 24, 28, 32, 36, 40, 44, 48, 52, 56, 60, 64, 68, 72, 76, 80, 84, 88, 92, 96, 100, 104, 108, 112, 116, 120, 124, 128, 132, 136, 140, 144, 148, 152, 156, 160, 164, 168, 172, 176, 180, 184, 188, 192, 196, 200], [5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80, 85, 90, 95, 100, 105, 110, 115, 120, 125, 130, 135, 140, 145, 150, 155, 160, 165, 170, 175,

In [27]:
va.test_basi_4(operazioni_py, operazioni_np)

In [ ]:
print("Misura del tempo di esecuzione dell'implementazione Python:")
%timeit operazioni_py()

print("Misura del tempo di esecuzione dell'implementazione NumPy:")
%timeit operazioni_np()

Consideriamo ora il *Setaccio di Eratostene*, un antico algoritmo per calcolare la tabella dei numeri primi fino a un certo numero $N$. Il procedimento è piuttosto semplice: si scrivono tutti i numeri naturali fino a $N$ in un elenco (il *setaccio*). Poi si cancellano tutti i multipli del primo numero del setaccio (iniziando da 2), escludendo il numero stesso. Si ripete poi l'operazione con il successivo numero non cancellato, fino a superare $\sqrt{N}$. Si può dimostrare che, per ogni numero $i$, la ricerca dei multipli da cancellare può iniziare da $i^2$, in quanto i multipli precedenti sono già stati cancellati.  
L'animazione nella figura seguente esemplifica il procedimento per $N=120$: la ricerca si può interrompere al numero 11, in quanto $\left \lfloor{\sqrt{N}}\right \rfloor = 10$.

![Setaccio di Eratostene](pynp/Eratostene.gif)

<img style="float:left" src="images/ar.png" /> **Esercizio 5** - Nella cella seguente completate la funzione `eratostene(n)` che deve implementare l'algoritmo appena descritto e restituire un array NumPy contenente tutti i numeri primi inferiori o uguali al parametro `n`. Si chiede di implementare il *setaccio* come un array NumPy di `n+1` booleani: al termine dell'algoritmo, il valore dell'elemento di indice `i` di tale array deve essere `True` se `i` è primo, `False` altrimenti. Gli elementi di indice 0 e 1 vanno impostati *manualmente* a `False`, per gli altri si deve utilizzare l'algoritmo. La funzione deve infine utilizzare tale array booleano per restituire un array NumPy di interi contenente i numeri primi individuati.  
Suggerimenti: per la radice quadrata si può utilizzare l'operatore \*\* (oppure importare il modulo math e utilizzare math.sqrt()), per l'intero inferiore si può utilizzare int() oppure math.floor()).  
Eseguire la cella seguente per controllare la funzione in alcuni casi di prova.

In [41]:
import math

def eratostene(n):
    
    # --- Svolgimento Esercizio 5: Inizio --- #
    
    mask = np.ones(n, dtype=bool)
    
    for i in range(2, math.floor(math.sqrt(n)) + 1):
        to_start = i*i
        mask[to_start::i] = False
        
    return mask
    
    # --- Svolgimento Esercizio 5: Fine --- #

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True])

In [42]:
va.test_basi_5(eratostene)

<img style="float:left" src="images/ar.png" /> **Esercizio 6** - Nella prossima cella completate la funzione `es6(n)`, che deve svolgere le seguenti operazioni attraverso opportune istruzioni Python e NumPy:

* Creare un array NumPy $p$ contenente i numeri primi minori o uguali a $2^n$ (a tale scopo utilizzare la funzione `eratostene()`;
* Create un array NumPy $d$ contenente le prime $n+1$ potenze del 2 {$2^0$, $2^1$, ..., $2^n$};
* Costruire una matrice $A$ tale che $A_{i,j} = min(p_i, d_j)$
* Restituire la somma degli elementi di $A$.

Suggerimenti: in NumPy è disponibile una *ufunc* che calcola il minimo elemento-per-elemento fra due matrici: come tutte le ufunc permette anche di utilizzare il broadcasting.  
Eseguire la cella seguente per controllare la funzione in alcuni casi di prova.

In [ ]:
def es6(n):
    
    # --- Svolgimento Esercizio 6: Inizio --- #    
    
    return -1

    # --- Svolgimento Esercizio 6: Fine --- #

In [ ]:
va.test_basi_6(es6)

Concludiamo questa esercitazione eseguendo l'ultima funzione che abbiamo creato con il parametro $n=3$; è sufficiente eseguire l'ultima cella. Questa potrebbe essere la risposta che qualcuno stava aspettando, o forse no: l'esercitazione è comunque conclusa.

In [ ]:
es6(3)